doc: Add release notes for 3.0.0 (#21339)
nrfconnect/sdk-nrf: doc/nrf/releases_and_maturity/migration/migration_guide_3.0.rst

Reviewer: Under `Protocols`, `Bluetooth Mesh` please add:

```
.. toggle::

   * Due to a incompatibility between the old and new ZMS backend for Settings, the mesh device will not be able to load its settings and provisioning data.
     This affects nRF54L devices.

     Make sure to unprovision mesh device before flashing the new firmware.

     Alternatively, enable the :kconfig:option:`CONFIG_SETTINGS_ZMS_LEGACY` Kconfig option to use old backed and recover the device settings and provisioning data.
```

Reviewer: * updated

Author: Added

The following changes are mandatory to make your application work in the same way as in previous releases.

nRF54H20
========
This section describes the changes specific to the nRF54H20 SoC and DK support in the |NCS|.

Dependencies
------------
The following required dependencies for the nRF54H20 SoC and DK have been updated.

nRF54H20 BICR
+++++++++++++

.. toggle::

   * The nRF54H20 BICR has been updated (from the one supporting |NCS| v2.9.0 as well as |NCS| v2.9.0-nRF54H20-1).
     To update the BICR of your development kit while in Root of Trust, do the following:

>      1. Build your application using |NCS| v2.9.0.

Reviewer 2: ```suggestion
     1. Build your application using |NCS| v3.0.0.
```

Author: Fixed

     #. Connect the nRF54H20 DK to your computer using the **DEBUGGER** port on the DK.

        .. note::

           On MacOS, connecting the DK might repeatedly trigger a popup displaying the message ``Disk Not Ejected Properly``.
           To disable this, run ``JLinkExe``, then run ``MSDDisable`` in the J-Link Commander interface.

     #. List all the connected development kits to see their serial number (matching the one on the DK's sticker):

        .. code-block::

           nrfutil device list

     #. Program the BICR by running nRF Util from your application folder using the following command:

        .. code-block::

           nrfutil device program --options chip_erase_mode=ERASE_NONE --firmware ./build/<your_application_name>/zephyr/bicr.hex --core Application --serial-number <serial_number>

nRF54H20 SoC binaries
+++++++++++++++++++++

.. toggle::

>    * The *nRF54H20 SoC binaries* bundle has been updated to version 0.9.5.

Reviewer 3: Should be 0.9.6 instead

Author: Fixed

     .. note::
        The nRF54H20 SoC binaries only support specific versions of the |NCS| and do not support rollbacks to a previous version.
        Upgrading the nRF54H20 SoC binaries on your development kit might break the DK's compatibility with applications developed for previous versions of the |NCS|.
        For more information, see :ref:`abi_compatibility`.

     To update the SoC binaries bundle of your development kit while in Root of Trust, do the following:

     1. Download the `nRF54H20 SoC binaries v0.9.6`_.

        .. note::
           On macOS, ensure that the ZIP file is not unpacked automatically upon download.

     #. Purge the device as follows:

        .. code-block::

           nrfutil device recover --core Application --serial-number <serial_number>
           nrfutil device recover --core Network --serial-number <serial_number>
           nrfutil device reset --reset-kind RESET_PIN --serial-number <serial_number>

     #. Run ``west update``.
     #. Move the correct :file:`.zip` bundle to a folder of your choice, then run nRF Util to program the binaries using the following command

        .. code-block::

>            nrfutil device x-suit-dfu --firmware nrf54h20_soc_binaries_v0.9.5.zip --serial-number <serial_number>

Reviewer 3: Should be 0.9.6 instead

Author: Fixed

     / {
        chosen {
           zephyr,entropy = &prng;
        };
        /delete-node/ psa-rng;
        prng: prng {
           compatible = "nordic,entropy-prng";
           status = "okay";
        };
     };
     /* Disable IPC between cpusec <-> cpurad. */
     &cpusec_cpurad_ipc {
        status = "disabled";
     };
     &cpurad_ram0x_region {
        status = "disabled";
     };
     &cpusec_bellboard {
        status = "disabled";
     };

SUIT MPI configuration
++++++++++++++++++++++

.. toggle::

   The SUIT MPI configuration has been moved from local Kconfig options to sysbuild.
   To migrate your application, move all ``CONFIG_MPI_*`` options from the application configuration into the :file:`sysbuild.conf` file.
   For example, to migrate the root manifest vendor ID, remove the following line from the :file:`prj.conf` file:

   .. code-block:: kconfig

      CONFIG_SUIT_MPI_ROOT_VENDOR_NAME="acme.corp"

   And add the following line inside the :file:`sysbuild.conf` file:

   .. code-block:: kconfig

      SB_CONFIG_SUIT_MPI_ROOT_VENDOR_NAME="acme.corp"

   If your project does not use the :file:`sysbuild.conf` file, you must create one.

> Samples and applications

Reviewer 4: Please add under `Samples and applications`:
```
nRF Desktop
-----------

.. toggle::

   * The default devices names (the :kconfig:option:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option) have been updated to remove the "52" infix.
     As a result of this change new peripherals won't pair with old dongles and the other way around.
     Also aligned the :file:`99-hid.rules` file inside the HID Configurator script.
     The HID Configurator rule won't work with old device names.
     To keep the backwards compatibility, do not update the kconfig:option:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option and "peer_name" array settings inside the :file:`ble_scan_def.h` files.
     To keep the HID Configurator rule working, make sure that "ATTRS{name}" field in the :file:`99-hid.rules` file is the same as the device Bluetooth name.
```

Reviewer 4: @divipillai, @MarekPieta please review.

Reviewer 5: ```
nRF Desktop
-----------

.. toggle::

   * The default devices names (the :kconfig:option:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option) have been updated to remove the "52" infix, because the nRF Desktop application supports also other SoC Series.
     As a result of this change, peripherals using firmware from NCS 3.0.0 (or newer) won't pair with dongles using firmware from an older NCS release and the other way around.
     Also aligned the :file:`99-hid.rules` file inside the HID Configurator script.
     The HID Configurator rule won't work with old device names.

     To keep backwards compatibility revert locally changes introduced by `commit_hash`:

     * :kconfig:option:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option defines device name used by HID peripheral.
     * "peer_name" array inside the :file:`ble_scan_def.h` file determines device name filters used by HID dongle while scanning for unpaired HID peripherals.
     * the :file:`99-hid.rules` file allows HID configurator Python script to configure nRF Desktop devices without root access.
```

btw. nRF Desktop app-specific Kconfigs need to use `:ref:` syntax (`:kconfig:option:` cannot be used for them)

Reviewer 4: Okay, so the final version after applying :ref: fix would be:
```
nRF Desktop
-----------

.. toggle::

   * The default devices names (the :ref:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option) have been updated to remove the "52" infix, because the nRF Desktop application supports also other SoC Series.
     As a result of this change, peripherals using firmware from NCS 3.0.0 (or newer) won't pair with dongles using firmware from an older NCS release and the other way around.
     Also aligned the :file:`99-hid.rules` file inside the HID Configurator script.
     The HID Configurator rule won't work with old device names.

     To keep backwards compatibility revert locally changes introduced by commit hash: ``5b80e46478462907a3cc4fd1686e241591775ffe``:

     * :ref:`CONFIG_DESKTOP_DEVICE_PRODUCT` Kconfig option defines device name used by HID peripheral.
     * "peer_name" array inside the :file:`ble_scan_def.h` file determines device name filters used by HID dongle while scanning for unpaired HID peripherals.
     * the :file:`99-hid.rules` file allows HID configurator Python script to configure nRF Desktop devices without root access.
 ```

Author: Added the content. Changed to :ref:`CONFIG_DESKTOP_DEVICE_PRODUCT <**config_desktop_app_options**>` to fix doc build error.

     1. Download the `nRF54H20 SoC binaries v0.9.6`_.

        .. note::
           On macOS, ensure that the ZIP file is not unpacked automatically upon download.

     #. Purge the device as follows:

        .. code-block::

           nrfutil device recover --core Application --serial-number <serial_number>
           nrfutil device recover --core Network --serial-number <serial_number>
           nrfutil device reset --reset-kind RESET_PIN --serial-number <serial_number>

     #. Run ``west update``.
     #. Move the correct :file:`.zip` bundle to a folder of your choice, then run nRF Util to program the binaries using the following command

        .. code-block::

           nrfutil device x-suit-dfu --firmware nrf54h20_soc_binaries_v0.9.6.zip --serial-number <serial_number>

     #. Purge the device again as follows:

        .. code-block::

           nrfutil device recover --core Application --serial-number <serial_number>
           nrfutil device recover --core Network --serial-number <serial_number>
           nrfutil device reset --reset-kind RESET_PIN --serial-number <serial_number>

> 

Reviewer 6: ```suggestion

SDK and toolchain
+++++++++++++++++

.. toggle::

  * To update the SDK and the toolchain, do the following:

   1. Open Toolchain Manager in nRF Connect for Desktop.
   #. Click :guilabel:`SETTINGS` in the navigation bar to specify where you want to install the |NCS|.
   #. In :guilabel:`SDK ENVIRONMENTS`, click the :guilabel:`Install` button next to the |NCS| version |release|.

nRF Util
++++++++

.. toggle::

  * ``nrfutil`` has been updated to v7.13.0.

    Install nRF Util v7.13.0 as follows:

      1. Download the nRF Util executable file from the `nRF Util development tool`_ product page.
      #. Add nRF Util to the system path on Linux and macOS, or environment variables on Windows, to run it from anywhere on the system.
         On Linux and macOS, use one of the following options:

         * Add nRF Util's directory to the system path.
         * Move the file to a directory in the system path.

      #. On macOS and Linux, give ``nrfutil`` execute permissions by typing ``chmod +x nrfutil`` in a terminal or using a file browser.
         This is typically a checkbox found under file properties.
      #. On macOS, to run the nRF Util executable, you need to allow it in the system settings.
      #. Verify the version of the nRF Util installation on your machine by running the following command::

            nrfutil --version

      #. If your version is below 7.13.0, run the following command to update nRF Util::

            nrfutil self-upgrade

         For more information, see the `nRF Util`_ documentation.

nRF Util device
+++++++++++++++

.. toggle::

  * nRF Util ``device`` command has been updated to v2.8.8

    Install the nRF Util ``device`` command v2.8.8 as follows::

       nrfutil install device=2.8.8 --force

    For more information, consult the `nRF Util`_ documentation.

nRF Util trace
++++++++++++++

.. toggle::

  * nRF Util ``trace`` command has been updated to v3.3.0.

    Install the nRF Util ``trace`` command v3.3.0 as follows::

       nrfutil install trace=3.3.0 --force

    For more information, consult the `nRF Util`_ documentation.

nRF Util suit
+++++++++++++

.. toggle::

  * nRF Util ``suit`` command has been updated to v0.9.0.

    Install the nRF Util ``suit`` command v0.9.0 as follows::

       nrfutil install suit=0.9.0 --force

    For more information, consult the `nRF Util`_ documentation.

```
I'd add also something about nRF util

Author: Added

Author: Added.

           };
        };
        /* Disable IPC between cpusec <-> cpurad. */
        &cpusec_cpurad_ipc {
           status = "disabled";
        };
        &cpurad_ram0x_region {
           status = "disabled";
        };
        &cpusec_bellboard {
           status = "disabled";
        };

SUIT MPI configuration
++++++++++++++++++++++

.. toggle::

   The SUIT MPI configuration has been moved from local Kconfig options to sysbuild.
   To migrate your application, move all ``CONFIG_MPI_*`` options from the application configuration into the :file:`sysbuild.conf` file.
   For example, to migrate the root manifest vendor ID, remove the following line from the :file:`prj.conf` file:

   .. code-block:: kconfig

      CONFIG_SUIT_MPI_ROOT_VENDOR_NAME="acme.corp"

   And add the following line inside the :file:`sysbuild.conf` file:

   .. code-block:: kconfig

      SB_CONFIG_SUIT_MPI_ROOT_VENDOR_NAME="acme.corp"

   If your project does not use the :file:`sysbuild.conf` file, you must create one.

Application development
=======================

The following are the changes required to migrate your applications to the |NCS| 3.0.0.

ZMS Settings backend

> --------------------

Author: @rghaddab, split and  moved the contents here as discussed offline.

     If you prefer to continue using ``nrfjprog`` for programming devices, :ref:`specify the west runner <programming_selecting_runner>` with ``west flash``.

   * Erasing the external memory when programming a new firmware image with ``west flash`` series now always correctly honors the ``--erase`` flag (and its absence) both when using the ``nrfjprog`` and ``nrfutil`` backends.
     Prior to this release, the ``nrjfprog`` backend would always erase only the sectors of the external flash used by the new firmware, and the ``nrfutil`` backend would always erase the whole external flash.

> 

Author: Added under Required changes

> Samples and applications

Reviewer 4: Please add following entry under `Samples and applications`:
```
Bluetooth Fast Pair samples
---------------------------

* :ref:`fast_pair_locator_tag` sample:
  Because for the ``nrf52840dk/nrf52840`` board target the MCUBoot DFU signature type has been changed to the Elliptic curve digital signatures with curve P-256 (ECDSA P256 - the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_ECDSA_P256`` Kconfig option), you will not be able to perform DFU from an old version to a new one.
  If you want to keep the DFU possibility, do not change the MCUBoot DFU signature type.
```

Reviewer 4: @peknis, @kapi-no, please review.

Reviewer 7: ```
Bluetooth Fast Pair samples
---------------------------

* :ref:`fast_pair_locator_tag` sample:
  If you want to align your application project with the newest version of the :ref:`fast_pair_locator_tag` sample and still maintain the DFU backwards compatibility for your already deployed products that are based on the ``nrf52840dk/nrf52840`` board target, please use the RSA signature algorithm (the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_RSA`` Kconfig option) that is supported as part of the previous |NCS| releases.
  In the current |NCS| release, the MCUboot DFU signature type for the ``nrf52840dk/nrf52840`` board target has been changed to the Elliptic curve digital signatures with curve P-256 (ECDSA P256 - the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_ECDSA_P256`` Kconfig option), and you will not be able to perform DFU from an old version to a new one.

Author: Added

   * Errors that were previously notified to the application with the ``LWM2M_CARRIER_ERROR_RUN`` event type have instead been added to :c:macro:`LWM2M_CARRIER_ERROR_CONFIGURATION`.

Bluetooth Fast Pair Locator tag sample
--------------------------------------

.. toggle::

>    If you want to align your application project with the newest version of the :ref:`fast_pair_locator_tag` sample and still maintain the DFU backwards compatibility for your already deployed products that are based on the ``nrf52840dk/nrf52840`` board target, use the RSA signature algorithm (the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_RSA`` Kconfig option) that is supported as part of the previous |NCS| releases.
>    In the current |NCS| release, the MCUboot DFU signature type for the ``nrf52840dk/nrf52840`` board target has been changed to the Elliptic curve digital signatures with curve P-256 (ECDSA P256 - the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_ECDSA_P256`` Kconfig option), and you will not be able to perform DFU from an old version to a new one.

Reviewer 7: ```suggestion
   If you want to align your application project with the newest version of the :ref:`fast_pair_locator_tag` sample and still maintain the DFU backwards compatibility for your already deployed products that are based on the ``nrf52840dk/nrf52840``  and the ``nrf54l15dk/nrf54l15/cpuapp`` board targets, use the RSA signature algorithm (the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_RSA`` Kconfig option) that is supported as part of the previous |NCS| releases.
   In the current |NCS| release, the MCUboot DFU signature type has been changed:

    * To the Elliptic curve digital signatures with curve P-256 (ECDSA P256 - the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_ECDSA_P256`` Kconfig option) in case of the ``nrf52840dk/nrf52840`` board target.
    * To the Edwards-curve digital signature with curve Curve25519 (ED25519 - the ``SB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519`` Kconfig option) in case of the ``nrf54l15dk/nrf54l15/cpuapp`` board target.
    
    As a result, you will not be able to perform DFU from an old version to a new one.
```

Author: Added

>      The factory-programmed Asset Tracker v2 firmware is still available to program the nRF91 Series devices using the `Programmer app`_, the `Quick Start app`_, and the `Cellular Monitor app`_.

Reviewer 8: ```suggestion
     The factory-programmed Asset Tracker v2 firmware is still available to program the nRF91xx DKs using the `Programmer app`_, the `Quick Start app`_, and the `Cellular Monitor app`_.
```
as requested by Martin.

        .. code-block::

           nrfutil install device=2.8.8 --force

     If you prefer to continue using ``nrfjprog`` for programming devices, :ref:`specify the west runner <programming_selecting_runner>` with ``west flash``.

   * Erasing the external memory when programming a new firmware image with the ``west flash`` series now always correctly honors the ``--erase`` flag (and its absence) both when using the ``nrfjprog`` and ``nrfutil`` backends.
     Before this release, the ``nrjfprog`` backend would always erase only the sectors of the external flash used by the new firmware, and the ``nrfutil`` backend would always erase the whole external flash.

Application development
=======================

The following are the changes recommended to migrate your applications to the |NCS| 3.0.0.

> Performance optimization for ZMS Settings backend

Author: @rghaddab, split and  moved the contents here as discussed offline.

   * Errors that were previously notified to the application with the ``LWM2M_CARRIER_ERROR_RUN`` event type have instead been added to :c:macro:`LWM2M_CARRIER_ERROR_CONFIGURATION`.

> 

Reviewer 9: For Wi-Fi,  please add:

Wi-Fi®
+------
+
+.. toggle::
+
+   * For samples using Wi-Fi features:
+
+      * nRF70 driver heap was part of the system shared heap ``CONFIG_HEAP_MEM_POOL_SIZE``. Now dedicated heaps have been defined for Wi-Fi driver control plane and data plane operations.
+        Default value of heap for control plane operations is 20000 bytes and for data plane operations, it is 130000. It is recommended to disable ``CONFIG_HEAP_MEM_POOL_IGNORE_MIN``,
+        and let system calculate the ``K_HEAP`` size. Any subsequent RAM overflow issues need
+        to be solved by fine tuning ``CONFIG_NRF_WIFI_CTRL_HEAP_SIZE`` and ``CONFIG_NRF_WIFI_DATA_HEAP_SIZE``.

Author: This will be added after the merge of https://github.com/nrfconnect/sdk-nrf/pull/21958.

Reviewer 9: The change mentioned here has already been merged. So, this may not be dependent on merging of #21958

Author: OK. Added.
